pkg/ebpf/netlink/consumer.go

```diff
@@ -0,0 +1,290 @@
+		if err := setSocketBufferSize(netlinkBufferSize, c.conn); err != nil {
+			log.Errorf("error setting rcv buffer size for netlink socket: %s", err)
+		}
+
+		if size, err := getSocketBufferSize(c.conn); err == nil {
+			log.Debugf("rcv buffer size for netlink socket is %d bytes", size)
+		}
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Consumer) Events() chan Event {
+	output := make(chan Event, outputBuffer)
+
+	c.do(false, func() {
+		c.conn.JoinGroup(netlinkCtNew)
+		c.receive(output, false)
+	})
+
+	return output
+}
+
+func (c *Consumer) DumpTable(family uint8) chan Event {
+	output := make(chan Event, outputBuffer)
+
+	c.do(false, func() {
+		defer close(output)
+
+		req := netlink.Message{
+			Header: netlink.Header{
+				Flags: netlink.Request | netlink.Dump,
+				Type:  netlink.HeaderType((unix.NFNL_SUBSYS_CTNETLINK << 8) | ipctnlMsgCtGet),
+			},
+			Data: []byte{family, unix.NFNETLINK_V0, 0, 0},
+		}
+
+		verify, err := c.conn.Send(req)
+		if err != nil {
+			output <- c.eventFor(nil, err)
+			return
+		}
+
+		if err := netlink.Validate(req, []netlink.Message{verify}); err != nil {
+			output <- c.eventFor(nil, err)
+			return
+		}
+
+		c.receive(output, true)
+	})
+
+	return output
+}
+
+func (c *Consumer) Stop() {
+	c.conn.Close()
+}
+
+func (c *Consumer) initWorker(procRoot string) {
+	go func() {
+		util.WithRootNS(procRoot, func() {
+			for {
+				fn, ok := <-c.workQueue
+				if !ok {
+					return
+				}
+				fn()
+			}
+		})
+	}()
+}
+
+func (c *Consumer) do(sync bool, fn func()) {
+	if !sync {
+		c.workQueue <- fn
+		return
+	}
+
+	done := make(chan struct{})
+	syncFn := func() {
+		fn()
+		close(done)
+	}
+	c.workQueue <- syncFn
+	<-done
+}
+
+// Based on of https://github.com/mdlayher/netlink/conn.go
+// In our context we don't care about multi-part messages in the sense
+// that is better to flush their parts to the output channel (and generate back-pressure)
+// instead of buffering the whole thing in memory. When dump is set to true we terminate
+// after processing the first multi-part message.
+func (c *Consumer) receive(output chan Event, dump bool) {
+	for {
+		c.pool.Reset()
+		msgs, err := c.socket.Receive()
+		if err != nil {
+			if !isEOF(err) {
+				output <- c.eventFor(nil, err)
+			}
+			return
+		}
+
+		for _, m := range msgs {
+			if err := checkMessage(m); err != nil {
+				output <- c.eventFor(nil, err)
+				return
+			}
+		}
+
+		multiPartDone := len(msgs) > 0 && msgs[len(msgs)-1].Header.Type == netlink.Done
+		if multiPartDone {
+			msgs = msgs[:len(msgs)-1]
+		}
+
+		output <- c.eventFor(msgs, nil)
+
+		// If we're doing a conntrack dump it means we are done after reading the multi-part message
+		if dump && multiPartDone {
+			return
+		}
+	}
+}
+
+func (c *Consumer) eventFor(msgs []netlink.Message, err error) Event {
+	return Event{
+		msgs:    msgs,
+		err:     err,
+		buffers: c.pool.inUse,
+		pool:    c.pool,
+	}
+}
+
+type bufferPool struct {
+	inUse [][]byte
```

Comment:
ah, I ideally I would like to have `socket.ReceiveInto([]byte)` method and wouldn't need to do that. but that's basically to be able to bind a `[]byte` buffer to an `[]netlink.Message`. I think I might actually be able to implement `socket.ReceiveInto([]byte)` since I no longer need to implement the `netlink.Socket` interface, which I needed when I started working on this.